### PR TITLE
[m3tsz] Fast path decoding for constant time series

### PR DESCRIPTION
**Pull request by linasm**
**What this PR does / why we need it**:
Implements  fast path decoding of constant datapoints which are encoded as bit sequences of `001` (with `intOptimized = true`) or `00` (with `intOptimized = false`).
Naturally, the resultant performance change is heavily dependent on the underlying data, so I have added new microbenchmarks to illustrate the results at the extreme ends of the spectrum.

Macrobenchmark:

- running `read_data_files` in datapoint benchmark mode on some real world data fileset:
```
read_data_files -b 1596009600000000000 -p ~/testdata/m3db -s 14 -B datapoints
```

- before the change:
```
Running time: 280.909914ms

7279 series read
(25912.22 series/second)

3821758 datapoints decoded
(13604923.89 datapoints/second)
```

- after the change:
```
Running time: 210.245028ms

7279 series read
(34621.51 series/second)

3821758 datapoints decoded
(18177637.95 datapoints/second)
```

Yielding a 25% improvement.

Microbenchmarks:

- before the change:
```
BenchmarkM3TSZDecode
BenchmarkM3TSZDecode-12            	   28538	     42534 ns/op
BenchmarkM3TSZDecodeConstant
BenchmarkM3TSZDecodeConstant-12    	   32270	     37174 ns/op
BenchmarkM3TSZDecodeVarying
BenchmarkM3TSZDecodeVarying-12     	   27801	     43690 ns/op
```

- after the change:
```
BenchmarkM3TSZDecode
BenchmarkM3TSZDecode-12            	   33792	     36631 ns/op
BenchmarkM3TSZDecodeConstant
BenchmarkM3TSZDecodeConstant-12    	   79882	     15064 ns/op
BenchmarkM3TSZDecodeVarying
BenchmarkM3TSZDecodeVarying-12     	   24435	     46400 ns/op
```

Best case (`BenchmarkM3TSZDecodeConstant`) yields ~60% improvement (from 37174 ns/op to 15064 ns/op).
Worst case (`BenchmarkM3TSZDecodeVarying`) yields ~6% performance loss (from 43690 ns/op to 46400 ns/op).
Mixed case (`BenchmarkM3TSZDecode` - 10 real world time series, few of them constant) yields ~14% improvement (from 42534 ns/op to 36631 ns/op).

### CODE REVIEW

File: src/dbnode/encoding/m3tsz/iterator.go [hunk 1]
```diff
@@ -83,6 +83,49 @@ func (it *readerIterator) Next() bool {
...
+			// 0: TimeEncodingScheme.ZeroBucket().Opcode()
+			// 0: opcodeZeroValueXOR
+			if it.peekBits(2) == 0b00 {
+				if it.err != nil {
+					return false
+				}
+
+				it.tsIterator.PrevTime += xtime.UnixNano(it.tsIterator.PrevTimeDelta)
+				it.curr.TimestampNanos = it.tsIterator.PrevTime
+				it.tsIterator.PrevAnt = nil
```
Comment (vpranckaitis) [inline src/dbnode/encoding/m3tsz/iterator.go:120]: `FloatEncoderAndIterator` seems to also be setting `PrevXor = 0` on `opcodeZeroValueXOR`

https://github.com/m3db/m3/blob/87f37951bf2274b6d5560ec595ce508ac4944fc1/src/dbnode/encoding/m3tsz/float_encoder_iterator.go#L123-L126

File: src/dbnode/encoding/m3tsz/roundtrip_test.go [hunk 2]
```diff
@@ -268,3 +275,15 @@ func generateOverflowDatapoints() []ts.Datapoint {
...
 	return res
 }
+
+func generateConstDatapoints(numPoints int) []ts.Datapoint {
+	currentTime := xtime.FromSeconds(1630436128)
+	res := make([]ts.Datapoint, 0, numPoints)
+
+	for i := 0; i < numPoints; i++ {
+		currentTime = currentTime.Add(time.Second)
+		res = append(res, ts.Datapoint{TimestampNanos: currentTime, Value: 5})
```
Comment (vpranckaitis) [inline src/dbnode/encoding/m3tsz/roundtrip_test.go:285]: Would it also be useful to have a test with constant non-integer value?

File: src/dbnode/encoding/m3tsz/iterator.go (hunk 1 above)
Comment (linasm) [reply src/dbnode/encoding/m3tsz/iterator.go:120]: Good catch. Funny thing though, I believe, this would be impossible to reproduce even in theory :)

Comment (linasm) [reply src/dbnode/encoding/m3tsz/iterator.go:120]: I mean, with using the encoder. Probably possible to craft such input manually.

**Review by vpranckaitis — APPROVED** [review]

**Pull request merged**